### PR TITLE
Promises support

### PR DESCRIPTION
**Pull request by piranna**
Add support for `fs.promises`. It also add support for the newer functions not available before. The pull-request is not bacward compatible due to removal of `SyncWriteStream` function (seems a typo for me since I was not able to find it in the docs, please confirm and I'll re-add it if needed), and fix resolution of up dir to work the same way as `..` direntry in root dir, that in POSIX systems it points to itself, so `/../blah` is the same as `/blah`.

Please tell me any concerns do you have :-)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,26 +1,45 @@
 var up = require('./lib/up');
 var slice = [].slice;
 
+
+const ONE_ARGUMENT_PROMISES =
+[
```
Comment (juliangruber) [inline index.js]: could you please remove those newlines?

Comment (juliangruber) [reply index.js]: let me add standard to this :)

**Review by juliangruber — CHANGES_REQUESTED** [review]

File: index.js (hunk 1 above)
Comment (juliangruber) [reply index.js]: Could you please run `npx prettier-standard '**/*.js'` on your patch? That should make it mergeable with current master.

Comment (piranna) [reply index.js]: Done :-)

**Comment by piranna** [comment]
There was still some conflicts, now they are fixed.

**Comment by juliangruber** [comment]
This looks great, thank you! Would you mind also adding some tests, so we can verify this doesn't break in the future? If not that is ok also and I can merge as is.

**Comment by piranna** [comment]
> This looks great, thank you!

You are welcome :-)


> Would you mind also adding some tests, so we can verify this doesn't break in the future? If not that is ok also and I can merge as is.

No problem. Can you be able to configure Travis CI and Coveralls while I'm adding them? :-)

**Comment by juliangruber** [comment]
Travis is set up already. Coveralls I've never done, if I add you as an owner on this repo, could you do it? And would you be interested in maintaining this in general?

**Comment by piranna** [comment]
> Travis is set up already

But... it's not checking the pull-requests :-/ Maybe you need to enable that feature?


> Coveralls I've never done, if I add you as an owner on this repo, could you do it?

Yes, I think there would not be problems.


> And would you be interested in maintaining this in general?

Maybe :-) Main problem is publishing to npm, besides that I would not have so much trouble having another child to feed :-D

**Comment by juliangruber** [comment]
> But... it's not checking the pull-requests :-/ Maybe you need to enable that feature?

It should run on the next commit, travis was disabled for some reason.

> Yes, I think there would not be problems.

Invited you :)

> Maybe :-) Main problem is publishing to npm, besides that I would not have so much trouble having another child to feed :-D

I can add you to both!

**Comment by piranna** [comment]
> > But... it's not checking the pull-requests :-/ Maybe you need to enable that feature?
> 
> It should run on the next commit, travis was disabled for some reason.
> 
> >

Nice.


> > Yes, I think there would not be problems.
> 
> Invited you :)

Thank you! :-)


> > Maybe :-) Main problem is publishing to npm, besides that I would not have so much trouble having another child to feed :-D
> 
> I can add you to both!

That would be cool :-)

By the way, [I have just enabled Coveralls for the repo](https://coveralls.io/github/juliangruber/subfs), just only last to generate coverage info and send it.

**Comment by piranna** [comment]
I have updated the code to use ES8 features and optimize the resolve algorythm. I've also changed the API to make the filesystem optional using the Node.js`fs` module as default value, and added a `_resolve()` function bound to the used base dir.

Note: tests are not passing because you've configured Travis to use Node.js 0.10, not Node.js 10 :-P

**Comment by juliangruber** [comment]
> Note: tests are not passing because you've configured Travis to use Node.js 0.10, not Node.js 10 :-P

Just change the travis.yml, we'll release a major after this

**Comment by piranna** [comment]
Done, tests are passing. Should we update the tests to use `jest` so we can have code coverage by free, or is it good enough just like how tests are at this moment?

**Comment by juliangruber** [comment]
do as you wish :) either is ok for me

**Comment by piranna** [comment]
Tests converted to Jest :-)

**Review by juliangruber — APPROVED** [review]
nice work!

**Comment by piranna** [comment]
> nice work!

Thank you! :-D

**Pull request merged**